File: src/api/savedViews.ts
import { client } from './client';
import type { AppFilters } from '../lib/appFilters';
import type { RetailContextUrlState } from '../lib/insightDeepLink';

export type SavedViewModule = 'hub' | 'focus' | 'agents' | 'management';

export interface SavedViewApiState {
  tab: SavedViewModule;
  period?: string | null;
  filters: AppFilters;
  section?: string | null;
  subtab?: string | null;
}

export interface SavedViewItem {
  id: number;
  module_id: SavedViewModule;
  name: string;
  state: SavedViewApiState;
  schema_version: 1;
  created_at: string;
  updated_at: string;
}

interface SavedViewListResponse { items: SavedViewItem[] }
interface SavedViewDeleteResponse { ok: boolean }

export function toSavedViewApiState(state: RetailContextUrlState): SavedViewApiState {
  if (state.tab === 'settings') throw new Error('Settings cannot be saved as a view');
  const base = {
    tab: state.tab,
    period: state.period,
    filters: {
      ...state.filters,
      magazin: [...state.filters.magazin],
      agent: [...state.filters.agent],
    },
  };
  if (state.tab === 'hub') return { ...base, section: state.hubSection ?? 'current' };
  if (state.tab === 'focus') return { ...base, section: state.campaignSection ?? 'incentive' };
  if (state.tab === 'agents') return { ...base, section: state.agentsSection ?? 'overview' };
  return { ...base, subtab: state.managementSubtab ?? 'asm' };
}

export function toRetailContextUrlState(state: SavedViewApiState): RetailContextUrlState {
  const base = {
    tab: state.tab,
    period: state.period ?? undefined,
    filters: {
      ...state.filters,
      magazin: [...state.filters.magazin],
      agent: [...state.filters.agent],
    },
  };
  if (state.tab === 'hub') {
    return { ...base, hubSection: state.section as 'current' | 'history' | 'visits' };
  }
  if (state.tab === 'focus') {
    return {
      ...base,
      campaignSection: state.section as 'incentive' | 'promo' | 'concurs' | 'premium' | 'focus',
    };
  }
  if (state.tab === 'agents') {
    return { ...base, agentsSection: state.section as 'overview' | 'grile' | 'analysis' };
  }
  return {
    ...base,
    managementSubtab: state.subtab as 'asm' | 'target-calculator' | 'salarii' | 'pnl',
  };
}

export async function listSavedViews(signal?: AbortSignal): Promise<SavedViewItem[]> {
  const { data } = await client.get<SavedViewListResponse>('/api/saved-views', { signal });
  return data.items;
}

export async function createSavedView(
  name: string,
  state: RetailContextUrlState,
): Promise<SavedViewItem> {
  const { data } = await client.post<SavedViewItem>('/api/saved-views', {
    name,
    state: toSavedViewApiState(state),
  });
  return data;
}

export async function deleteSavedView(viewId: number): Promise<void> {
  const { data } = await client.delete<SavedViewDeleteResponse>(`/api/saved-views/${viewId}`);
  if (!data.ok) throw new Error('Saved view delete was not acknowledged');
}

File: backend/schemas/saved_views.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, field_validator, model_validator

from schemas.common import MonthStr, StrictApiModel

MAX_FILTER_VALUE_LENGTH = 180
MAX_SELECTION_ITEMS = 50
SavedViewModule = Literal["hub", "focus", "agents", "management"]


def _normalized_text(value: str, *, maximum: int) -> str:
    normalized = value.strip()
    if not normalized or len(normalized) > maximum or any(not char.isprintable() for char in normalized):
        raise ValueError("value must be non-blank printable text within bounds")
    return normalized


class SavedViewFilters(StrictApiModel):
    firma: str = Field(max_length=MAX_FILTER_VALUE_LENGTH)
    rm: str = Field(max_length=MAX_FILTER_VALUE_LENGTH)
    magazin: list[str] = Field(default_factory=list, max_length=MAX_SELECTION_ITEMS)
    agent: list[str] = Field(default_factory=list, max_length=MAX_SELECTION_ITEMS)

    @field_validator("firma", "rm")
    @classmethod
    def normalize_scalar(cls, value: str) -> str:
        return _normalized_text(value, maximum=MAX_FILTER_VALUE_LENGTH)

    @field_validator("magazin", "agent")
    @classmethod
    def normalize_selection(cls, values: list[str]) -> list[str]:
        result: list[str] = []
        seen: set[str] = set()
        for raw in values:
            value = _normalized_text(raw, maximum=MAX_FILTER_VALUE_LENGTH)
            if value in seen:
                continue
            seen.add(value)
            result.append(value)
        return result


class SavedViewState(StrictApiModel):
    tab: SavedViewModule
    period: MonthStr | None = None
    filters: SavedViewFilters
    section: str | None = Field(default=None, max_length=40)
    subtab: str | None = Field(default=None, max_length=40)

    @field_validator("section", "subtab")
    @classmethod
    def normalize_optional_text(cls, value: str | None) -> str | None:
        return None if value is None else _normalized_text(value, maximum=40)

    @model_validator(mode="after")
    def validate_module_context(self):
        if self.tab == "hub":
            if self.section not in {"current", "history", "visits"} or self.subtab is not None:
                raise ValueError("hub saved views require a valid section and no subtab")
        elif self.tab == "focus":
            if self.section not in {"incentive", "promo", "concurs", "premium", "focus"} or self.subtab is not None:
                raise ValueError("focus saved views require a valid section and no subtab")
        elif self.tab == "agents":
            if self.section not in {"overview", "grile", "analysis"} or self.subtab is not None:
                raise ValueError("agents saved views require a valid section and no subtab")
        elif self.section is not None or self.subtab not in {"asm", "target-calculator", "salarii", "pnl"}:
            raise ValueError("management saved views require a valid subtab and no section")
        return self


class SavedViewCreate(StrictApiModel):
    name: str = Field(min_length=1, max_length=80)
    state: SavedViewState

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        return _normalized_text(value, maximum=80)


class SavedViewUpdate(StrictApiModel):
    name: str | None = Field(default=None, max_length=80)
    state: SavedViewState | None = None

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str | None) -> str | None:
        return None if value is None else _normalized_text(value, maximum=80)

    @model_validator(mode="after")
    def require_change(self):
        if self.name is None and self.state is None:
            raise ValueError("at least one saved view field must change")
        return self


class SavedViewItem(StrictApiModel):
    id: int = Field(ge=1)
    module_id: SavedViewModule
    name: str
    state: SavedViewState
    schema_version: Literal[1]
    created_at: str
    updated_at: str


class SavedViewListResponse(StrictApiModel):
    items: list[SavedViewItem]


class SavedViewDeleteResponse(StrictApiModel):
    ok: bool
